### ISSUE TITLE
latest update causes PHP fatal error

### ISSUE THREAD

**Issue by cybermcm** [bitnami/bitnami-docker-dokuwiki#78]
**Description**

**Steps to reproduce the issue:**

1. trying to start container (data was there with an older bitnami version

**Describe the results you received:**

error
```
"GET /doku.php?id=debian_server HTTP/1.1" 500 -
[php7:error] [pid 93] [client 172.20.1.1:49246] PHP Fatal error:  Uncaught Error: Class 'dokuwiki\\plugin\\config\\core\\Setting\\Setting' not found in /opt/bitnami/dokuwiki/inc/deprecated.php:61\nStack trace:\n#0 /opt/bitnami/dokuwiki/inc/load.php(38): require_once()\n#1 /opt/bitnami/dokuwiki/inc/init.php(200): require_once('/opt/bitnami/do...')\n#2 /opt/bitnami/dokuwiki/doku.php(36): require_once('/opt/bitnami/do...')\n#3 {main}\n  thrown in /opt/bitnami/dokuwiki/inc/deprecated.php on line 61

```

**Describe the results you expected:**

working container

**Additional information you deem important (e.g. issue happens only occasionally):**

standard options set, nothing special, volume mounted '/bitnami/dokuwiki'

**Version**

- Output of `docker version`:

```
Version:           19.03.12
```

- Output of `docker info`:

```
root@xxx:/opt# docker info
Client:
 Debug Mode: false

Server:
 Containers: 65
  Running: 65
  Paused: 0
  Stopped: 0
 Images: 87
 Server Version: 19.03.12
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 7ad184331fa3e55e52b890ea95e65ba581ae3429
 runc version: dc9208a3303feef5b3839f4323d9beb36df0a9dd
 init version: fec3683
 Security Options:
  apparmor
  seccomp
   Profile: default
 Kernel Version: 4.19.0-10-amd64
 Operating System: Debian GNU/Linux 10 (buster)
 OSType: linux
 Architecture: x86_64
 CPUs: 2
 Total Memory: 11.73GiB
 Name: xxx
 ID: UIJG:UTZ2:NYOF:GVQS:DD2H:BTHD:JKA6:GXMN:M6JD:WNWT:GJJZ:I45E
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.1:2345
  127.0.0.0/8
 Live Restore Enabled: false

WARNING: No swap limit support
WARNING: bridge-nf-call-iptables is disabled
WARNING: bridge-nf-call-ip6tables is disabled

```

- Output of `docker-compose version` (if applicable):

```
root@xxx:/opt# docker-compose version
docker-compose version 1.26.2, build eefe0d31
docker-py version: 4.2.2
CPython version: 3.7.7
OpenSSL version: OpenSSL 1.1.0l  10 Sep 2019

```

**Comment by beltran-rubo** [comment]
Hi,

Note that Dokuwiki has moved to a major version. For Dokuwiki it is not possible to separate the persistent data & the application code itself so basically the whole application is stored in the volume folder and do not support automated updates.

Please check the upgrade process from Dokuwiki itself to upgrade the application files on your own https://www.dokuwiki.org/install:upgrade

**Comment by sugarat333** [comment]
I too have just had my Dokuwiki instance explode when running the newer container.  I thought the value-add in using Bitnami was that upgrades would be executed when pulling a newer image.  It seems that we're on our own with a broken wiki now, because Dokuwiki has upgraded their version.  If we have to do this all ourselves, I'm really not sure what the point of using Bitnami was.  :-(

**Comment by Aephir** [comment]
Has anyone successfully made the upgrade using this docker image, who would be willing to share a "how to"? I'm guessing clicking the "upgrade" from the prompt in the UI wouldn't work?

<img width="1176" alt="Screenshot 2020-08-06 at 16 22 48" src="https://user-images.githubusercontent.com/20152788/89543394-28d78000-d801-11ea-9cce-79313500d96f.png">

But then following the general guide @beltran-rubo linked to, is it really "just" copying all files from the unpacked tarball and overwriting **everything**? And how did you handle permissions of folders and files?

I also guess that it means that this will be a recurring issue at every new major version of dokuwiki? So perhaps dokuwiki just isn't really that well suited for Docker (at least, not a lot gained)? Or am I missing something?

**Comment by MrReSc** [comment]
Unfortunately I did not manage to update my DokuWiki either. 
Can someone please tell me how to fix this?

**Comment by cybermcm** [comment]
I downloaded the source files here [https://download.dokuwiki.org/](url) , just stable and no languages/plugins (sufficient in my case)
I unzipped it and just copied (overwrite) the files that where already present, in my case conf/data/lib directories. Within these directories I only copied the existing folders, not all...
Following up with a chown on all files (user 1001)
Dokuwiki now works again but  I consider this a really "dirty" upgrade. It was simple in my case as I didn't use any plugin/extension. It may be more difficult for others.
I'll report back if there is any problem appearing...

**Comment by beltran-rubo** [comment]
Right, the process is not ideal but Dokuwiki, as far as we know, do not have a clear separation in different folders of the main logic and other files like themes or plugins. That is why the recommended upgrade process from Dokuwiki is just overwrite the files.

It is not possible to separate the logic (container) from the data (volume) so it can not be upgraded with simply replacing the container images.

**Comment by Aephir** [comment]
Hmmm, seems like I'll run an old version for now, and look at this in a week or two, once I've om paternity leave :)

In the meantime, I've posted a suggestion on the dokuwiki forum to split out the code a bit cleaner, to allow this (https://forum.dokuwiki.org/d/18150-re-organizing-structure-of-application-to-allow-containerized-updates). Feel free to chime in if you have any comments, suggestions, or corrections.

@cybermcm I'm a little bit unclear about the technical details on this, but when they say that "user 1001 is the Apache daemon" ([here](https://github.com/bitnami/bitnami-docker-dokuwiki#020180422202005011246-debian-10-r68)), that's only true within the docker container, right? So it's 1001, even if 1001 is used for a different user on my base OS, because that's the user ID within the container, and that's what needs access to the dirs/files?

**Comment by cybermcm** [comment]
@Aephir: You summed it up correctly. User 1001 is from within the docker container. This user (inside the container) has to access your persistent volume files on your host, so you have to set up user permissions for user 1001 on your host (easiest way is to chown the whole dokuwiki directory to user 1001)

**Comment by marafa** [comment]
would love to see a step by step instructions to get this working again

if i understand correctly. save the data directory, delete the lib and conf directories and start over again? that does not work for me. it just loops and then fails

```
dokuwiki 00:28:40.77 
dokuwiki 00:28:40.77 Welcome to the Bitnami dokuwiki container
dokuwiki 00:28:40.77 Subscribe to project updates by watching https://github.com/b
dokuwiki 00:28:40.77 Submit issues and feature requests at https://github.com/bitn
dokuwiki 00:28:40.77 
dokuwiki 00:28:40.78 INFO  ==> ** Starting DokuWiki setup **
dokuwiki 00:28:40.81 INFO  ==> Configuring PHP options
dokuwiki 00:28:40.87 INFO  ==> Restoring persisted DokuWiki installation
realpath: /bitnami/dokuwiki/lib/images/smileys/local: No such file or directory
```

**Comment by cybermcm** [comment]
@marafa: do not delete anything in the first place. Download the new package and just upload and overwrite the existing directories, in my case conf and data and lib folder. Then set the permissions to user 1001 (chown all files recursively).
If everything is working you can delete the unused files according to this page: [https://www.dokuwiki.org/install:unused_files](url) , in my case only 3 files.
hope this helps

**Comment by marcosbc** [comment]
@marafa Currently the following directories are persisted and must be copied to `/bitnami/dokuwiki` for the initialization to work properly:

* `data`
* `conf`
* `lib/plugins`
* `lib/tpl`
* `lib/images/smileys/local`
* `lib/images/interwiki`

What you can do is temporarily set `command: sleep infinity` in your `docker-compose.yml`, and then copy those directories from a DokuWiki tarball to the volume at `/bitnami/dokuwiki`. I.e. for `lib/tpl` it must be located at `/bitnami/dokuwiki/lib/tpl`.

If your DokuWiki volume is mounted on your file system, instead of using Docker volumes, then it's even easier: Just extract the files to the proper locations and it should work.

**Comment by Aephir** [comment]
I've run into issues using this approach. 

First with a plugin ([rssticker](https://www.dokuwiki.org/plugin:rssticker)) gave issues, and the container would not even start. But simply removing the directory `sudo rm -rf dokuwiki_testing/lib/plugins/rssticker/` and restarting the container somewhat worked.

But now all the buttons at the top (`user tools`, `add new page` and `profile`) are unresponsive. And I'm not sure if this is due to the brute force removing a plugin directory (that is "in use" in the wiki), or something else? Another plugin interfering?

If I try to click the drop-down menu for `user tools`, `add new page`, or `profile`, respectively, I get three to four lines in the log. Not sure if they are helpful to anyone, but certainly not to me. And nothing happens in the UI.

```
172.18.0.1 - - [11/Aug/2020:10:50:01 +0000] "GET /doku.php?id=start HTTP/1.1" 200 10000
172.18.0.1 - - [11/Aug/2020:10:50:01 +0000] "GET /lib/exe/taskrunner.php?id=start&1597143001 HTTP/1.1" 200 42
172.18.0.1 - - [11/Aug/2020:10:50:01 +0000] "GET /lib/exe/fetch.php?media=wiki:logo.png HTTP/1.1" 304 -
172.18.0.1 - - [11/Aug/2020:10:50:02 +0000] "GET /lib/exe/manifest.php HTTP/1.1" 200 334
172.18.0.1 - - [11/Aug/2020:10:50:02 +0000] "GET /lib/exe/fetch.php?media=wiki:dokuwiki.svg HTTP/1.1" 304 -
172.18.0.1 - - [11/Aug/2020:10:50:02 +0000] "GET /doku.php?id=start HTTP/1.1" 200 10000
172.18.0.1 - - [11/Aug/2020:10:50:03 +0000] "GET /lib/exe/taskrunner.php?id=start&1597143002 HTTP/1.1" 200 42
172.18.0.1 - - [11/Aug/2020:10:50:03 +0000] "GET /lib/exe/fetch.php?media=wiki:logo.png HTTP/1.1" 304 -
172.18.0.1 - - [11/Aug/2020:10:50:03 +0000] "GET /lib/exe/manifest.php HTTP/1.1" 200 334
172.18.0.1 - - [11/Aug/2020:10:50:03 +0000] "GET /lib/exe/fetch.php?media=wiki:dokuwiki.svg HTTP/1.1" 304 -
```

Any ideas of how to get it the last few steps of the way to an upgrade?

**Comment by marcosbc** [comment]
@Aephir Unfortunately I don't know what may be going on, you mentioned it was unresponsive but there are many things that could mean (i.e. redirection to a closed port, JS not working, PHP request timeout...). Have you checked the ["Troubleshooting" section](https://www.dokuwiki.org/install:upgrade#troubleshooting
) in the DokuWiki upgrade page? It looks like something that could be somehow related to caching.

If that did not work, we would recommend to try and migrate only the data (`/bitnami/dokuwiki/data`) to the new release, without overwriting `lib` and `conf` from the new deployment.

Since most DokuWiki data is stored there, and does not contain any code, it should be fully compatible between versions.
Afterwards, you would just need to adapt your configuration and add the plugins from the old release manually.

**Comment by marafa** [comment]
so!
what i did was backup my work and started from scratch
what i saw was the directories and files now had ownership of `1001:root` at runtime but then i saw 
```
dokuwiki 21:59:42.76 INFO  ==> Configuring file permissions for DokuWiki
```
after which the ownership turned to `1001:777`

ALSO

the file permissions required are 644 at the beginning but after that "fixing" it was as such
```
drwxrwxr-x.  2 1001 777 4096 Aug 12 18:07 conf
drwxrwxr-x. 12 1001 777 4096 Aug 12 18:07 data
drwxr-xr-x.  5 1001 777   46 Aug 12 18:07 lib
```

**Comment by daveserpa** [comment]
For those having difficulties, I successfully upgraded my installation with the following procedure:

1. Backup the volume mounted at /bitnami/dokuwiki.
2. Rollback container to bitnami/dokuwiki:20180422.
3. Install and run @splitbrain's [Upgrade Plugin](https://github.com/splitbrain/dokuwiki-plugin-upgrade/).
4. Upgrade container to bitnami/dokuwiki:20200729.

**Comment by marcosbc** [comment]
Thanks for sharing! Hopefully this will be enough to upgrade any installation to a new release.

> what i saw was the directories and files now had ownership of `1001:root` at runtime but then i saw
>```dokuwiki 21:59:42.76 INFO  ==> Configuring file permissions for DokuWiki```
>after which the ownership turned to `1001:777`

@marafa I'm not sure how that could have happened. In my case, the permissions are set to `1001:root` for files in `/bitnami/dokuwiki`, and `daemon:root` for files in `/opt/bitnami/dokuwiki`. In both cases the group is set to `root`. In any case, I assume this does not happen anymore (since it is working)?

**Comment by Aephir** [comment]
Yes, thanks @daveserpa , certainly easier this way. 

However, it didn't quite do it for me. @marcosbc put me on the right path, but  actually ended up using the upgrade plugin, just with a few modifications, and it at least seems to work now.

I had some issues with plugins. First, the [rssticker](https://www.dokuwiki.org/plugin:rssticker) completely broke everything (even when installing on a clean "hogfather", the web interface just disappeared). So my approach, in case anyone else runs into issues, starting with just running your old image:

1. Remove rssticker.
2. Remove all plugins marked as [deprecated] (a couple for me, though not sure if that is needed)
3. Upgrade all upgradable plugins (I suspect some of them were updated for compatibility with new dokuwiki release)
4. Run upgrade plugin, as per @daveserpa 's post
5. Upgrade container, again as per Dave's post (though I used the `bitnami/dokuwiki:latest`; time will tell if I'm brave or dumb)
